### PR TITLE
Add support for EXECUTE IMMEDIATE statement

### LINKED ISSUE #17353: Add support for EXECUTE IMMEDIATE statement
Currently, running a prepared statement requires 3 requests: `PREPARE`, `EXECUTE... USING`, and `DEALLOCATE`.  In some cases, once a statement is prepared, it's executed many times before it's deallocated, so the cost of additional requests gets amortized.  However, some frameworks will generate a sequence like above for every operation.

Proposal:

Add a new parametrized statement that also includes the parameters to be bound.  Note that while the SQL spec contains a statement with similar syntax and semantics, we are deviating from the standard (as it does not allow parameters to be bound).

Syntax:
```
EXECUTE IMMEDIATE
'SELECT * FROM t WHERE a = ? and b = ?'
USING 'foo', 42
```

The semantics are the same as:
```
PREPARE stmt1 FROM SELECT * FROM t WHERE a = ? and b = ?
EXECUTE stmt1 USING 'foo', 42
DEALLOCATE PREPARE stmt1
```

### PR DESCRIPTION
**Pull request by aalbu**
## Description

Fixes #17353 

Syntax:
```
EXECUTE IMMEDIATE
'SELECT * FROM t WHERE a = ? and b = ?'
USING 'foo', 42
```

The semantics are the same as:
```
PREPARE stmt1 FROM SELECT * FROM t WHERE a = ? and b = ?
EXECUTE stmt1 USING 'foo', 42
DEALLOCATE PREPARE stmt1
```

## Release notes

( ) This is not user-visible or docs only and no release notes are required.
(X) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

**Comment by findepi** [comment]
I posed some syntax comments in the issue: https://github.com/trinodb/trino/issues/17353

**Review by electrum — APPROVED** [review]
Looks good. The last two commits should be squashed.

@martint should also review

File: core/trino-main/src/main/java/io/trino/execution/QueryPreparer.java [hunk 1]
```diff
@@ -65,6 +68,19 @@ public PreparedQuery prepareQuery(Session session, Statement wrappedStatement)
             prepareSql = Optional.of(session.getPreparedStatementFromExecute(executeStatement));
             statement = sqlParser.createStatement(prepareSql.get(), createParsingOptions(session));
         }
+        else if (statement instanceof ExecuteImmediate executeImmediateStatement) {
+            // HACK: pad the embedded statement with white space, so that it starts at the same position
+            //       as in the original EXECUTE IMMEDIATE statement, in order to get meaningful parse errors
+            int line = executeImmediateStatement.getStatementLocation().map(NodeLocation::getLineNumber).orElse(1);
+            int column = executeImmediateStatement.getStatementLocation().map(NodeLocation::getColumnNumber).orElse(1);
+            String prefix = "\n".repeat(line - 1) + " ".repeat(column); // account for the starting single quote
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/execution/QueryPreparer.java:76]: Hmm... this is not great. We can deal with the parsing errors by catching the exception and rewriting the line/column.

I need to think about how to handle semantic or execution errors, though.

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/jdbc/TestPreparedStatements.java [hunk 2]
```diff
@@ -120,6 +120,24 @@ public void preparedSelectSql()
         }
     }
 
+    @Test(groups = JDBC)
+    @Requires(ImmutableAllTypesTable.class)
+    public void executeImmediateSelectSql()
+            throws SQLException
+    {
+        final int testValue = 2147483647;
```
Comment (martint) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/jdbc/TestPreparedStatements.java:128]: Remove `final`

File: core/trino-main/src/main/java/io/trino/execution/QueryPreparer.java (hunk 1 above)
Comment (aalbu) [reply core/trino-main/src/main/java/io/trino/execution/QueryPreparer.java:76]: Well, I'm not very proud of this, but I couldn't figure out a way to handle it when catching the exception.  It's not obvious how to compute the correct position when the embedded SQL statement spans more than one line.  Also, this hack seems to handle semantic exceptions, as well (see the test in my fixup commit).

Comment (electrum) [reply core/trino-main/src/main/java/io/trino/execution/QueryPreparer.java:76]: Are you fine with this approach for now?

Comment (martint) [reply core/trino-main/src/main/java/io/trino/execution/QueryPreparer.java:76]: It’s probably ok for now, but you might want to check for downstream quirks such as how the query is being logged, for instance.

Also, add a TODO to fix it. A more robust approach would be to handle this all within the analyzer by having dedicated logic to deal with the ExecuteImmediate node there instead of doing it inside QueryPreparer.

Comment (aalbu) [reply core/trino-main/src/main/java/io/trino/execution/QueryPreparer.java]: I have added a TODO.  I tried to add a test showing that the original query text is preserved in the `QueryCompletedEvent`, but I was not able to get event listeners invoked in tests (`QueryMonitor#queryCompletedEvent` does not get invoked by the test runners).  I was able to confirm through a manual test that the original query text is preserved.

File: core/trino-main/src/main/java/io/trino/execution/QueryPreparer.java [hunk 3]
```diff
@@ -61,23 +64,42 @@ public PreparedQuery prepareQuery(Session session, Statement wrappedStatement)
...
+            prepareSql = Optional.of(session.getPreparedStatementFromExecute(executeStatement));
             statement = sqlParser.createStatement(prepareSql.get(), createParsingOptions(session));
         }
-
-        if (statement instanceof ExplainAnalyze) {
-            Statement innerStatement = ((ExplainAnalyze) statement).getStatement();
+        else if (statement instanceof ExecuteImmediate executeImmediateStatement) {
+            // TODO: find a non-hacky way to get correct position information in case of syntactic and semantic errors,
+            //       perhaps handle this all within the analyzer by having dedicated logic to deal with the ExecuteImmediate
+            //       node there instead of doing it in this class
```
Comment (kasiafi) [inline core/trino-main/src/main/java/io/trino/execution/QueryPreparer.java:74]: I dealt with the same kind of issue when parsing JSON path within a query. Maybe you find it helpful: https://github.com/trinodb/trino/blob/71e2a87a6a4bef0df7ec9d717a669abe74e9d6fb/core/trino-parser/src/main/java/io/trino/sql/jsonpath/PathParser.java#L43

**Review by kasiafi — COMMENTED** [review]
just skimmed

File: docs/src/main/sphinx/language/reserved.rst [hunk 4]
```diff
@@ -47,6 +47,7 @@ Keyword                        SQL:2016      SQL-92
 ``GROUP``                      reserved      reserved
 ``GROUPING``                   reserved
 ``HAVING``                     reserved      reserved
+``IMMEDIATE``                  reserved      reserved
```
[the remark this answers is not in the document]
Comment (aalbu) [reply docs/src/main/sphinx/language/reserved.rst:50]: Thanks for catching that.  I don't have access to the 2016 standard, but I saw that in [SQL-92](http://www.contrib.andrew.cmu.edu/~shadow/sql/sql1992.txt), it was a reserved word, so I assumed it continued to be.

File: core/trino-main/src/main/java/io/trino/execution/QueryPreparer.java [hunk 5]
```diff
@@ -65,19 +68,39 @@ public PreparedQuery prepareQuery(Session session, Statement wrappedStatement)
...
-        if (statement instanceof ExplainAnalyze explainAnalyzeStatement) {
+        else if (statement instanceof ExecuteImmediate executeImmediateStatement) {
+            // TODO: consider adding dedicated logic to deal with ExecuteImmediate in the analyzer
+            try {
+                statement = sqlParser.createStatement(executeImmediateStatement.getStatement(), createParsingOptions(session));
+            }
+            catch (ParsingException e) {
+                int embeddedStatementLine = executeImmediateStatement.getStatementLocation().getLineNumber();
+                int embeddedStatementColumn = executeImmediateStatement.getStatementLocation().getColumnNumber();
+                ParsingException exceptionWithUpdatedLocation = new ParsingException(
```
[the remark this answers is not in the document]
Comment (martint) [reply core/trino-main/src/main/java/io/trino/execution/QueryPreparer.java:79]: Also, this won't work for a test such as this, as it only handles parsing errors:

```
        assertTrinoExceptionThrownBy(() -> assertions.query("EXECUTE IMMEDIATE 'SELECT * FROM tiny.tpch.orders'"))
                .hasMessageMatching("line 1:34: Catalog 'tiny' does not exist");
```

[the remark this answers is not in the document]
Comment (kasiafi) [reply core/trino-main/src/main/java/io/trino/execution/QueryPreparer.java:79]: If the SqlParser could take a "relative start position", it would solve both issues:
- parsing exceptions would have the correct location,
- the created AST would have the correct locations, so the Analyzer errors would be positioned correctly, too.

[the remark this answers is not in the document]
Comment (kasiafi) [reply core/trino-main/src/main/java/io/trino/execution/QueryPreparer.java:79]: >Also, this won't work for a test such as this, as it only handles parsing errors:

BTW @martint we have the same problem with `EXECUTE`:
```
trino> prepare test from select * from foo;
PREPARE
trino> execute test;
Query 20230510_184850_00010_fczhd failed: line 3:3: Schema must be specified when session schema is not set
```
It's even worse, as the line/column in error message refer to how the statement was formatter by the SqlFormatter, not to how it was originally written by the user.

Probably in this case we should skip the position altogether, or place it at the beginning of the `EXECUTE` statement, or maybe include the executed statement in the message to gie context for the error.

File: core/trino-parser/src/main/java/io/trino/sql/parser/SqlParser.java [hunk 6]
```diff
@@ -84,6 +86,11 @@ public Statement createStatement(String sql, ParsingOptions parsingOptions)
         return (Statement) invokeParser("statement", sql, SqlBaseParser::singleStatement, parsingOptions);
     }
 
+    public Statement createStatement(String sql, NodeLocation relativeLocation, ParsingOptions parsingOptions)
```
Comment (martint) [inline core/trino-parser/src/main/java/io/trino/sql/parser/SqlParser.java:89]: Rename `relativeLocation` to just `location`.

File: core/trino-main/src/main/java/io/trino/execution/QueryPreparer.java [hunk 7]
```diff
@@ -65,19 +69,44 @@ public PreparedQuery prepareQuery(Session session, Statement wrappedStatement)
...
+                throw new TrinoException(
+                        SYNTAX_ERROR,
+                        new ParsingException(
+                                e.getErrorMessage(),
+                                (RecognitionException) e.getCause(),
+                                e.getLineNumber() + embeddedStatementLine - 1,
+                                e.getColumnNumber() + (embeddedStatementLine == 1 ? embeddedStatementColumn : 0))
+                                .getMessage());
+            }
+        }
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/execution/QueryPreparer.java:92]: This should not be needed anymore. It should be taken care of by the use of createStatement with a base location above.

**Review by martint — CHANGES_REQUESTED** [review]

File: core/trino-parser/src/main/java/io/trino/sql/parser/AstBuilder.java [hunk 8]
```diff
@@ -319,10 +320,12 @@
 {
     private int parameterPosition;
     private final ParsingOptions parsingOptions;
+    private final Optional<NodeLocation> relativeLocation;
 
-    AstBuilder(ParsingOptions parsingOptions)
+    AstBuilder(Optional<NodeLocation> relativeLocation, ParsingOptions parsingOptions)
```
[the remark this answers is not in the document]
Comment (aalbu) [reply core/trino-parser/src/main/java/io/trino/sql/parser/AstBuilder.java:325]: I thought about that, but I hesitated given that nothing was using it and there were be no tests exercising it.

File: core/trino-main/src/main/java/io/trino/execution/QueryPreparer.java (hunk 7 above)
Comment (aalbu) [reply core/trino-main/src/main/java/io/trino/execution/QueryPreparer.java:92]: This is needed to handle syntax errors.  This is why we would ideally make the lexer aware of the relative parsing position.

Comment (martint) [reply core/trino-main/src/main/java/io/trino/execution/QueryPreparer.java:92]: That should be handled in the SqlParser class, though.

Comment (aalbu) [reply core/trino-main/src/main/java/io/trino/execution/QueryPreparer.java:92]: Oh, good point.

File: core/trino-parser/src/main/java/io/trino/sql/parser/AstBuilder.java (hunk 8 above)
Comment (martint) [inline core/trino-parser/src/main/java/io/trino/sql/parser/AstBuilder.java:325]: Call this `location` or `baseLocation`

File: core/trino-parser/src/main/java/io/trino/sql/parser/AstBuilder.java [hunk 9]
```diff
@@ -940,6 +943,15 @@ public Node visitExecute(SqlBaseParser.ExecuteContext context)
                 visit(context.expression(), Expression.class));
     }
 
+    @Override
+    public Node visitExecuteImmediate(SqlBaseParser.ExecuteImmediateContext context)
```
Comment (martint) [inline core/trino-parser/src/main/java/io/trino/sql/parser/AstBuilder.java:947]: This belongs in the next commit

File: core/trino-parser/src/main/java/io/trino/sql/parser/SqlParser.java [hunk 10]
```diff
@@ -149,8 +161,20 @@ public Token recoverInline(Parser recognizer)
                 parser.getInterpreter().setPredictionMode(PredictionMode.LL);
                 tree = parseFunction.apply(parser);
             }
+            catch (ParsingException e) {
+                location.ifPresent(statementLocation -> {
+                    int embeddedStatementLine = statementLocation.getLineNumber();
+                    int embeddedStatementColumn = statementLocation.getColumnNumber();
```
Comment (martint) [inline core/trino-parser/src/main/java/io/trino/sql/parser/SqlParser.java:167]: `embeddedStatementLine` -> `line`
`embeddedStatementColumn` -> `column`

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/jdbc/TestPreparedStatements.java [hunk 11]
```diff
@@ -105,7 +105,7 @@ public void preparedSelectSql()
             throws SQLException
     {
         String prepareSql = "PREPARE ps1 from SELECT c_int FROM " + TABLE_NAME + " WHERE c_int = ?";
-        final int testValue = 2147483647;
+        int testValue = 2147483647;
```
Comment (martint) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/jdbc/TestPreparedStatements.java:108]: Unrelated change. It should be in a separate commit.

File: core/trino-parser/src/main/java/io/trino/sql/tree/ExecuteImmediate.java [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+        ExecuteImmediate o = (ExecuteImmediate) obj;
+        return Objects.equals(statement, o.statement) &&
+                Objects.equals(parameters, o.parameters);
+    }
+
+    @Override
+    public String toString()
+    {
+        return toStringHelper(this)
+                .add("dynamicSql", statement)
```
Comment (martint) [inline core/trino-parser/src/main/java/io/trino/sql/tree/ExecuteImmediate.java:84]: Rename dynamicSql -> statement

File: core/trino-parser/src/main/java/io/trino/sql/tree/ExecuteImmediate.java [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+    @Override
+    public <R, C> R accept(AstVisitor<R, C> visitor, C context)
+    {
+        return visitor.visitExecuteImmediate(this, context);
+    }
+
+    @Override
+    public List<? extends Node> getChildren()
+    {
+        return parameters;
```
Comment (martint) [inline core/trino-parser/src/main/java/io/trino/sql/tree/ExecuteImmediate.java:57]: This should include `statement`

File: core/trino-parser/src/main/java/io/trino/sql/tree/ExecuteImmediate.java [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+
+public class ExecuteImmediate
+        extends Statement
+{
+    private final StringLiteral statement;
+    private final List<Expression> parameters;
+
+    public ExecuteImmediate(NodeLocation location, StringLiteral statement, List<Expression> parameters)
+    {
+        super(Optional.of(requireNonNull(location, "location is null")));
```
Comment (martint) [inline core/trino-parser/src/main/java/io/trino/sql/tree/ExecuteImmediate.java:33]: requireNonNull not needed for this case. `Optional.of()` will fail with an NPE in such case.

File: core/trino-parser/src/main/java/io/trino/sql/tree/ExecuteImmediate.java [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+        this.statement = requireNonNull(statement, "statement is null");
+        this.parameters = ImmutableList.copyOf(requireNonNull(parameters, "parameters is null"));
```
Comment (martint) [inline core/trino-parser/src/main/java/io/trino/sql/tree/ExecuteImmediate.java:35]: Same here. `ImmutableList.copyOf` will fail with NPE if parameters is null, so the check is unnecessary.

File: core/trino-parser/src/main/java/io/trino/sql/tree/ExecuteImmediate.java (hunk 13 above)
Comment (aalbu) [reply core/trino-parser/src/main/java/io/trino/sql/tree/ExecuteImmediate.java]: Thanks

**Comment by colebow** [comment]
I really don't like that we're merging new SQL syntax without docs the day before our normal release day...

@aalbu could you please try to follow-up with a docs PR? I think you can likely just add a blurb to execute.rst
cc @martint

**Comment by aalbu** [comment]
> @aalbu could you please try to follow-up with a docs PR? I think you can likely just add a blurb to execute.rst cc @martint

Sorry, my bad.  I'll write something up.
